File: lib/test/t_IdentityMatrix_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file for the IdentityMatrix class
 *
 *  Copyright 2005-2021 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"

using namespace OT;
using namespace OT::Test;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);

  IdentityMatrix matrix1(2);
  matrix1.setName("matrix1");
  fullprint << "matrix1 = " << matrix1 << std::endl;

  Point pt ;
  pt.add(5.) ;
  pt.add(0.) ;
  fullprint << "pt = " << pt << std::endl;

  Point result ;
  result = matrix1.solveLinearSystem(pt);
  fullprint << "result = " << result << std::endl;

  Scalar determinant ;
  determinant = matrix1.computeDeterminant();
  fullprint << "determinant = " << determinant << std::endl;

  Collection<Scalar> ev(2);
  ev = matrix1.computeEigenValues().getCollection();
  fullprint << "ev = " << ev << std::endl;

  bool isSPD = matrix1.isPositiveDefinite();
  fullprint << "isSPD = " << isSPD << std::endl;

  SquareMatrix matrix2 = matrix1.computeCholesky();
  fullprint << "matrix2 = " << matrix2 << std::endl;

  return ExitCode::Success;
}

File: python/test/t_IdentityMatrix_std.py
#! /usr/bin/env python

import openturns as ot

ot.TESTPREAMBLE()

matrix1 = ot.IdentityMatrix(2)
print("matrix1 = ", repr(matrix1))

pt = ot.Point()
pt.add(5.0)
pt.add(0.0)
print("pt = ", repr(pt))

result = ot.Point()
result = matrix1.solveLinearSystem(pt)
print("result = ", repr(result))

determinant = matrix1.computeDeterminant()
print("determinant = %.6f" % determinant)

ev = ot.ScalarCollection(2)
ev = matrix1.computeEigenValues()
print("ev = ", repr(ev))

if matrix1.isPositiveDefinite():
    isSPD = "true"
else:
    isSPD = "false"
print("isSPD = ", isSPD)

matrix2 = matrix1.computeCholesky()
print("matrix2 = ", repr(matrix2))

det, sgn = matrix1.computeLogAbsoluteDeterminant()
print(det)
assert det == 0.0, "wrong det"

sv = matrix1.computeSingularValues()
print(sv)
assert sv == [1.0] * 2, "wrong sv"

ev, u, vT = matrix1.computeSVD()
print(ev, u, vT)
assert ev == [1.0] * 2, "wrong svd"

ev, v = matrix1.computeEV()
print(ev, v)
assert ev == [1.0] * 2, "wrong ev"

tr = matrix1.transpose()
print(tr)
assert tr == matrix1, "transpose"

cov = ot.CovarianceMatrix(2)
prod = matrix1 @ cov
print(prod)
assert prod == matrix1, "product /cov"
